Add new variable tidb_opt_three_stage_distinct_agg (#10419)
pingcap/docs: system-variables.md

- When the value is set to `OFF`, the non-transactional DML statement stops immediately at the first error and returns the error. All the following batches are canceled.
- When the value is set to `ON` and an error occurs in a batch, the following batches will continue to be executed until all batches are executed. All errors occurred during the execution process are returned together in the result.

> ### tidb_opt_3stage_distinct_agg <span class="version-mark">New in v6.3.0</span>

Reviewer: 变量名还能改吗？3stage 连在一起不太清晰，建议 3-stage 或 three-stage

Author: 可以的。那就 `tidb_opt_multi_stage_distinct_agg`吧，考虑到以后可能不仅仅局限于3阶段。

Reviewer: LGTM

> ### tidb_opt_3stage_distinct_agg <span class="version-mark">New in v6.3.0</span>

Reviewer: ```suggestion
### tidb_opt_three_stage_distinct_agg <span class="version-mark">New in v6.3.0</span>
```

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Default value: `ON`
- This variable specifies whether to rewrite a `COUNT(DISTINCT)` aggregate into 3 stages aggregation in MPP mode.

> - This variable currently applies to aggregation with only 1 `COUNT(DISTINCT)`.

Reviewer: ```suggestion
- This variable currently applies to aggregations with only one `COUNT(DISTINCT)`.
```

- Default value: `OFF`
- This variable sets whether the optimizer rewrites the aggregate functions with `DISTINCT` to the two-level aggregate functions, such as rewriting `SELECT b, COUNT(DISTINCT a) FROM t GROUP BY b` to `SELECT b, COUNT(a) FROM (SELECT b, a FROM t GROUP BY b, a) t GROUP BY b`. When the aggregation column has serious skew and the `DISTINCT` column has many different values, this rewriting can avoid the data skew in the query execution and improve the query performance.

### tidb_opt_three_stage_distinct_agg <span class="version-mark">New in v6.3.0</span>

- Scope: SESSION | GLOBAL

> - Persists to cluster: Yes

Reviewer 2: YES

Author: @shichun-0415 There are multiple uses: YES vs Yes.
Which one is preferred?

Reviewer: `Yes` is better. I will change `YES` to `Yes`.

Reviewer: Fixed in ddb9f3e

> - Default value: `ON`

Reviewer: ```suggestion
- Type: Boolean
- Default value: `ON`
```

- Default value: `OFF`
- This variable sets whether the optimizer rewrites the aggregate functions with `DISTINCT` to the two-level aggregate functions, such as rewriting `SELECT b, COUNT(DISTINCT a) FROM t GROUP BY b` to `SELECT b, COUNT(a) FROM (SELECT b, a FROM t GROUP BY b, a) t GROUP BY b`. When the aggregation column has serious skew and the `DISTINCT` column has many different values, this rewriting can avoid the data skew in the query execution and improve the query performance.

### tidb_opt_three_stage_distinct_agg <span class="version-mark">New in v6.3.0</span>

- Scope: SESSION | GLOBAL
- Persists to cluster: Yes
- Type: Boolean
- Default value: `ON`

> - This variable specifies whether to rewrite a `COUNT(DISTINCT)` aggregation into a three-stage aggregation in MPP mode.

Reviewer 3: ```suggestion
- This variable specifies whether to rewrite a `COUNT(DISTINCT)` aggregation condition in a query into a three-stage aggregation in the MPP mode.
```

> - This variable currently applies to aggregations with only one `COUNT(DISTINCT)`.

Reviewer 3: This is a confusing. Does it mean:
```suggestion
- This variable currently applies to the aggregation that only contains one `COUNT(DISTINCT)`.
```
